# Week3 — VectorRAG 외부 DB 구축 및 검증

2주차에서 정제한 `rag_chunks.jsonl`(37 chunk)을 **pgvector** 외부 DB에 적재하고,
유사도 검색이 올바르게 동작하는지 **정량·정성 검증**합니다.

기술 스택은 **bsvibe-app** 의 실제 임베딩 스택을 그대로 옮겼습니다.

| 영역 | 구현 | bsvibe 원본 |
| --- | --- | --- |
| 임베딩 | `litellm.aembedding` (bge-m3 via Ollama (1024d)) | `backend/embedding/provider.py` |
| 저장/검색 | SQLAlchemy async + asyncpg + pgvector raw SQL | `backend/embedding/storage/pg.py` |
| 코사인 | `embedding <=> CAST(:qv AS vector)`, similarity=`1-distance` | 동상 |

**검증 항목**
1. 임베딩 차원(1536) · 모델 일치 — DB `vector(1536)` 컬럼 ↔ 모델 차원 ↔ 행에 stamp된 dimension
2. 적재 개수 — chunk 수(37) == `SELECT count(*)`
3. 유사도 검색 — golden 15문항 top-K + 코사인 점수, 상식성 검토


In [ ]:
import asyncio, json
from pathlib import Path
import sys

import nest_asyncio  # 노트북 이벤트 루프에서 async 실행
nest_asyncio.apply()

from dotenv import load_dotenv
from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession, create_async_engine

ROOT = Path.cwd().parents[2] if (Path.cwd().name == "notebooks") else Path.cwd()
# 노트북은 assignments/week3-rag-db/notebooks 에서 실행 → repo root 까지 3단계
REPO_ROOT = Path.cwd().resolve()
while not (REPO_ROOT / ".git").exists() and REPO_ROOT != REPO_ROOT.parent:
    REPO_ROOT = REPO_ROOT.parent
SCRIPTS = REPO_ROOT / "assignments" / "week3-rag-db" / "scripts"
sys.path.insert(0, str(SCRIPTS))

load_dotenv(REPO_ROOT / ".env")
from vector_store import EMBED_DIM, EMBED_MODEL, RagVectorStore, embed_texts, load_chunks

DSN = "postgresql+asyncpg://rag:rag@localhost:5433/rag_week3"
CHUNKS = REPO_ROOT / "assignments/week2-rag-dataset/data/processed/rag_chunks.jsonl"
GOLDEN = REPO_ROOT / "assignments/week2-rag-dataset/data/eval/retrieval_golden.jsonl"
engine = create_async_engine(DSN)
print("embedding model:", EMBED_MODEL, "| expected dim:", EMBED_DIM)


18:20:45 - LiteLLM:WARNING: common_utils.py:979 - litellm: could not pre-load bedrock-runtime response stream shape — Bedrock event-stream decoding will be unavailable. Error: No module named 'botocore'


18:20:46 - LiteLLM:WARNING: common_utils.py:24 - litellm: could not pre-load sagemaker-runtime response stream shape — SageMaker event-stream decoding will be unavailable. Error: No module named 'botocore'


embedding model: text-embedding-3-small | expected dim: 1536


## 검증 ① — 임베딩 차원 및 모델 일치

DB `embedding` 컬럼의 `vector(N)` 차원, 행에 stamp된 `dimension`/`embedding_model`, 그리고 실제 쿼리 임베딩 길이가 모두 **1536** 으로 일치하는지 확인합니다.

In [ ]:
async def verify_dim():
    async with AsyncSession(engine) as s:
        # pgvector 는 vector(N) 의 N 을 atttypmod 에 그대로 저장
        col_dim = (await s.execute(text('''
            SELECT a.atttypmod FROM pg_attribute a
            JOIN pg_class c ON a.attrelid = c.oid
            WHERE c.relname = 'rag_chunks' AND a.attname = 'embedding'
        '''))).scalar()
        rows = (await s.execute(text(
            'SELECT DISTINCT embedding_model, dimension FROM rag_chunks'
        ))).mappings().all()
    return col_dim, rows

col_dim, model_rows = asyncio.get_event_loop().run_until_complete(verify_dim())
# 실제 쿼리 임베딩 차원
q_emb = asyncio.get_event_loop().run_until_complete(embed_texts(["dimension probe"]))[0]

print("DB vector() 컬럼 차원      :", col_dim)
print("행에 stamp된 (model, dim)  :", [dict(r) for r in model_rows])
print("실제 쿼리 임베딩 길이       :", len(q_emb))
print()
assert col_dim == EMBED_DIM, "컬럼 차원 불일치"
assert all(r["dimension"] == EMBED_DIM for r in model_rows), "stamp 차원 불일치"
assert all(r["embedding_model"] == EMBED_MODEL for r in model_rows), "모델 불일치"
assert len(q_emb) == EMBED_DIM, "쿼리 임베딩 차원 불일치"
print(f"✅ 차원·모델 일치: vector({col_dim}) == dimension {EMBED_DIM} == len(query_emb) {len(q_emb)}, model={EMBED_MODEL}")


DB vector() 컬럼 차원      : 1024
행에 stamp된 (model, dim)  : [{'embedding_model': 'ollama/bge-m3', 'dimension': 1024}]
실제 쿼리 임베딩 길이       : 1024

✅ 차원·모델 일치: vector(1024) == dimension 1024 == len(query_emb) 1024, model=ollama/bge-m3


## 검증 ② — 적재 데이터 개수(Count)

텍스트 스플리터로 쪼갠 chunk 수와 DB에 적재된 행 수가 일치하는지 확인합니다.

In [ ]:
chunks = load_chunks(CHUNKS)

async def db_count():
    async with AsyncSession(engine) as s:
        return await RagVectorStore(s).count()

n_db = asyncio.get_event_loop().run_until_complete(db_count())
print("입력 chunk 수 :", len(chunks))
print("DB 적재 count :", n_db)
assert len(chunks) == n_db, "count 불일치"
print(f"\n✅ count 일치: {len(chunks)} chunks == {n_db} rows")


입력 chunk 수 : 37
DB 적재 count : 37

✅ count 일치: 37 chunks == 37 rows


## 검증 ③ — 유사도 검색 (golden 15문항)

2주차 검색 평가셋(질문 ↔ 정답 source)을 그대로 사용합니다. 각 질문을 임베딩해 top-5 코사인 검색을 수행하고, **랭킹 정확도(Hit@1/@3, MRR)** 와 **top-1 코사인 점수** 를 봅니다.

In [ ]:
golden = [json.loads(l) for l in GOLDEN.read_text(encoding="utf-8").splitlines() if l.strip()]

async def run_golden():
    out = []
    async with AsyncSession(engine) as s:
        store = RagVectorStore(s)
        for g in golden:
            emb = (await embed_texts([g["question"]]))[0]
            hits = await store.search(query=emb, limit=5)
            ranked = [h.source_id for h in hits]
            expected = set(g["expected_source_ids"])
            rank = next((i + 1 for i, sid in enumerate(ranked) if sid in expected), None)
            out.append({
                "qid": g["qid"], "question": g["question"][:34],
                "top1_sim": round(hits[0].similarity, 3) if hits else None,
                "hit@1": bool(ranked[:1] and ranked[0] in expected),
                "hit@3": any(sid in expected for sid in ranked[:3]),
                "rank": rank,
            })
    return out

res = asyncio.get_event_loop().run_until_complete(run_golden())

import statistics as st
hit1 = sum(r["hit@1"] for r in res) / len(res)
hit3 = sum(r["hit@3"] for r in res) / len(res)
mrr = sum((1 / r["rank"]) if r["rank"] else 0 for r in res) / len(res)
sims = [r["top1_sim"] for r in res if r["top1_sim"] is not None]

print(f"{'qid':6} {'top1_sim':>8} {'hit@1':>6} {'hit@3':>6} {'rank':>5}  question")
for r in res:
    print(f"{r['qid']:6} {r['top1_sim']:>8} {str(r['hit@1']):>6} {str(r['hit@3']):>6} {str(r['rank']):>5}  {r['question']}")
print()
print(f"Hit@1 = {hit1:.1%}   Hit@3 = {hit3:.1%}   MRR = {mrr:.3f}")
print(f"top1 코사인: min {min(sims):.3f} / mean {st.mean(sims):.3f} / max {max(sims):.3f}")


qid    top1_sim  hit@1  hit@3  rank  question
q-001     0.607   True   True     1  BSVibe 모노레포는 어떤 구성으로 돌아가나요?
q-002     0.598   True   True     1  workspace 단위로 vault와 retriever를 어떻
q-003     0.583   True   True     1  이미 결정된 내용을 다음 run에서 재사용하려면 어디를 봐야 
q-004     0.534   True   True     1  파운더가 거절한 접근(부정 사례)은 어디에 저장되고 어떻게 검
q-005     0.643   True   True     1  canonical concept, decision, negat
q-006     0.688   True   True     1  agent runtime은 workspace retriever
q-007     0.621   True   True     1  settle 단계에서 vault 노트를 어떻게 흡수하고 임베딩
q-008     0.554   True   True     1  ingest 시 chunk 크기 한도를 어떻게 잡나요?
q-009     0.636   True   True     1  semantic 검색은 어떤 컬럼/인덱스로 동작하나요?
q-010     0.541   True   True     1  파운더가 잘못된 노드를 retract하면 어떤 흐름으로 처리되
q-011      0.63   True   True     1  undo 30초 카운트다운은 어디서 어떻게 계산되나요?
q-012     0.498  False   True     3  garden 관찰이 canonical concept으로 승격되
q-013     0.537   True   True     1  alias collision이나 redirect cycle 같
q-014     0.544   True   True     1  

### 정성 분석 — 코사인 절대값과 cross-lingual 갭

golden 질문은 **한국어**, 문서 본문은 **영어**입니다. 랭킹은 정확하지만 절대 코사인이 0.7 기준보다 낮게 나오는 경향이 있어, 동일 질문을 영어로도 던져 비교합니다.

In [ ]:
pair = {
    "KO (원본)": "workspace 단위로 vault와 retriever를 어떻게 묶나요?",
    "EN (동일 의미)": "How does KnowledgeFactory bind vault and retriever per workspace?",
}
async def compare():
    async with AsyncSession(engine) as s:
        store = RagVectorStore(s)
        for lang, q in pair.items():
            emb = (await embed_texts([q]))[0]
            h = (await store.search(query=emb, limit=1))[0]
            print(f"{lang:14} top1_sim={h.similarity:.3f}  src={h.source_id}")
asyncio.get_event_loop().run_until_complete(compare())
print()
print("→ 같은 정답 문서인데 EN 질의가 KO 질의보다 코사인이 높습니다 (cross-lingual 갭).")
print("→ 따라서 '0.7 이상' 절대 기준은 monolingual(EN↔EN) 가정. KO↔EN 에서는")
print("  랭킹 기반 지표(Hit@k/MRR)가 검색 품질의 더 신뢰할 신호입니다.")


KO (원본)        top1_sim=0.598  src=bsvibe-src-002
EN (동일 의미)     top1_sim=0.727  src=bsvibe-src-002

→ 같은 정답 문서인데 EN 질의가 KO 질의보다 코사인이 높습니다 (cross-lingual 갭).
→ 따라서 '0.7 이상' 절대 기준은 monolingual(EN↔EN) 가정. KO↔EN 에서는
  랭킹 기반 지표(Hit@k/MRR)가 검색 품질의 더 신뢰할 신호입니다.


### 검증 ④ — sparse baseline (BM25) 비교

dense pgvector 가 Hit@3=100% 라는 절대값만 보면 "충분히 좋다" 인지 가늠 어렵습니다. 동일 데이터셋·동일 평가셋에 **순수 stdlib BM25** (k1=1.5, b=0.75, doc = `text + title + tags`) 를 sparse baseline 으로 돌려서 dense 가 얼마나 끌어올렸는지 정량으로 봅니다.

`scripts/bm25_baseline.py` 를 그대로 import 해 같은 `evaluate()` 를 호출합니다.


In [ ]:
from bm25_baseline import load_chunks_as_docs, evaluate, CHUNKS_PATH, GOLDEN_PATH

sparse = evaluate(load_chunks_as_docs(CHUNKS_PATH), GOLDEN_PATH)
dense = {"hit@1": hit1, "hit@3": hit3, "mrr": mrr}

print(f"{'metric':6} {'sparse':>8} {'dense':>8} {'gain':>10}")
for m, dv in dense.items():
    sv = sparse[m]
    suffix = "%p" if m.startswith("hit") else ""
    print(f"{m:6} {sv:>8.3f} {dv:>8.3f}  +{dv - sv:.3f}{suffix}")
print()
print(f"sparse 상세: Hit@1={sparse['hit@1']:.1%}, Hit@3={sparse['hit@3']:.1%}, MRR={sparse['mrr']:.3f}")


metric   sparse    dense       gain
hit@1     0.600    0.933  +0.333%p
hit@3     0.733    1.000  +0.267%p
mrr       0.692    0.956  +0.264

sparse 상세: Hit@1=60.0%, Hit@3=73.3%, MRR=0.692


**해석**

1. **코드/식별자 데이터셋이라 BM25 도 만만치 않다** — `pgvector`, `settle`, `frontmatter`, `canonical` 같은 영문 키워드가 한국어 질문에 그대로 섞여 lexical 매칭이 잘 됨 (sparse 만으로 Hit@3 73%).
2. **dense 가 메우는 것** — sparse 가 못 잡은 케이스는 한영 혼합 토큰(`retract하면`)처럼 형태소 분리가 필요한 질문, 또는 의미는 같지만 어휘가 다른 케이스 (`negative pattern` ↔ `거절한 접근`). dense embedding 이 의미 유사도로 메워 Hit@3 100% 달성.
3. **다음 단계 비교 기준** — hybrid (sparse + dense) 나 rerank 도입 시 두 지표가 floor + ceiling 으로 직접 비교에 쓰임.


## 검증 ⑤ — 어떤 질문에 어떤 top-5 가 나오는가

aggregate metric (Hit@k, MRR) 만 보면 모델이 어떤 *방식* 으로 맞히고 틀리는지 안 보입니다. 대표 3가지 질문에 대해 sparse(BM25) 와 dense(pgvector) 의 실제 top-5 source_id + 점수 + 정답 여부 (✓) 를 같이 봅니다.

- `q-001`: 단순 매칭 — 둘 다 1위
- `q-007`: 멀티 정답 — 둘 다 top-2 안에 정답 2건 회수
- `q-010`: cross-lingual 한계 — sparse 가 lexical 만으로 못 잡는 케이스


In [ ]:
from bm25_baseline import (
    bm25_ranked_source_ids,
    load_chunks_as_docs,
    tokenize,
    CHUNKS_PATH,
    GOLDEN_PATH,
)

_preview = {}
for _line in CHUNKS_PATH.read_text(encoding="utf-8").splitlines():
    if _line.strip():
        _d = json.loads(_line)
        _sid = _d["source_id"]
        if _sid not in _preview:
            _preview[_sid] = _d["metadata"].get("title", "")[:48]

_docs = load_chunks_as_docs(CHUNKS_PATH)
_golden = {g["qid"]: g for g in [
    json.loads(l) for l in GOLDEN_PATH.read_text(encoding="utf-8").splitlines() if l.strip()
]}
PICKS = ["q-001", "q-007", "q-010"]

for qid in PICKS:
    g = _golden[qid]
    expected = set(g["expected_source_ids"])
    ranked = bm25_ranked_source_ids(_docs, tokenize(g["question"]))[:5]
    print(f"[{qid}] {g['question']}")
    print(f"  expected: {sorted(expected)}")
    print(f"  sparse(BM25) top-5:")
    for i, (sid, score) in enumerate(ranked, 1):
        mark = " ✓" if sid in expected else "  "
        print(f"    {i}. {sid:18}  score={score:5.2f}{mark}  {_preview.get(sid, '')}")
    print()


[q-001] BSVibe 모노레포는 어떤 구성으로 돌아가나요?
  expected: ['bsvibe-src-001']
  sparse(BM25) top-5:
    1. bsvibe-src-001    score= 0.03 ✓  BSVibe monorepo layout and local stack
    2. bsvibe-src-030    score= 0.02    compose.prod.yaml overrides dev defaults for pro
    3. bsvibe-src-004    score= 0.02    Resolved decisions are retrieved from settled ga
    4. bsvibe-vault-001  score= 0.02    Fact note frontmatter — typed triple with supers
    5. bsvibe-src-031    score= 0.02    Deploy README documents the migrate-on-boot and 

[q-007] settle 단계에서 vault 노트를 어떻게 흡수하고 임베딩하나요?
  expected: ['bsvibe-src-008', 'bsvibe-src-025']
  sparse(BM25) top-5:
    1. bsvibe-src-008    score= 5.24 ✓  Settle worker writes vault notes, promotes conce
    2. bsvibe-src-025    score= 4.37 ✓  Settle runtime factories build per-settlement ex
    3. bsvibe-src-015    score= 2.91    Cross-run decision reuse is tested through seed 
    4. bsvibe-src-016    score= 1.96    RetractionService orchestrates ontology correcti
 

In [ ]:
async def _dense_top5(question: str):
    emb = (await embed_texts([question]))[0]
    async with AsyncSession(engine) as s:
        store = RagVectorStore(s)
        hits = await store.search(query=emb, limit=5)
    return [(h.source_id, h.similarity) for h in hits]

for qid in PICKS:
    g = _golden[qid]
    expected = set(g["expected_source_ids"])
    ranked = asyncio.get_event_loop().run_until_complete(_dense_top5(g["question"]))
    print(f"[{qid}] {g['question']}")
    print(f"  expected: {sorted(expected)}")
    print(f"  dense(pgvector) top-5:")
    for i, (sid, sim) in enumerate(ranked, 1):
        mark = " ✓" if sid in expected else "  "
        print(f"    {i}. {sid:18}  sim={sim:5.3f}{mark}  {_preview.get(sid, '')}")
    print()


[q-001] BSVibe 모노레포는 어떤 구성으로 돌아가나요?
  expected: ['bsvibe-src-001']
  dense(pgvector) top-5:
    1. bsvibe-src-001    sim=0.607 ✓  BSVibe monorepo layout and local stack
    2. bsvibe-src-030    sim=0.441    compose.prod.yaml overrides dev defaults for pro
    3. bsvibe-src-004    sim=0.419    Resolved decisions are retrieved from settled ga
    4. bsvibe-src-006    sim=0.385    Composite retriever merges multiple knowledge so
    5. bsvibe-src-031    sim=0.381    Deploy README documents the migrate-on-boot and 

[q-007] settle 단계에서 vault 노트를 어떻게 흡수하고 임베딩하나요?
  expected: ['bsvibe-src-008', 'bsvibe-src-025']
  dense(pgvector) top-5:
    1. bsvibe-src-025    sim=0.621 ✓  Settle runtime factories build per-settlement ex
    2. bsvibe-src-008    sim=0.596 ✓  Settle worker writes vault notes, promotes conce
    3. bsvibe-src-015    sim=0.494    Cross-run decision reuse is tested through seed 
    4. bsvibe-src-023    sim=0.478    EventBus enumerates lifecycle, vault, and ingest
    5. bsvibe

**관찰**

- `q-001` — sparse 1위 score 0.03 vs 2위 0.02 로 마진 거의 없음 (운에 가까운 1위). dense 는 sim 0.607 로 회수, 차순위(0.441) 와 0.17 마진 → 안정적.
- `q-007` — 영문 식별자 (`settle`, `vault`) 가 섞여 sparse 도 top-2 에 정답 2건 깔끔히 회수. dense 도 동일하게 top-2 에 회수 (sim 0.621 / 0.596) — 일치.
- `q-010` — **결정적 케이스**. `retract하면` 같은 한영 혼합 토큰이 tokenizer 에서 분리 안 돼 sparse 가 top-5 모두 score=0.00 → lexical 로 풀 수 없음. dense 는 `RetractionSignal` (0.541) 을 **rank 1** 에, `RetractionService` (0.486) 를 rank 5 에 회수 → expected 3건 중 2건 hit. cross-lingual + 어휘 다양성 영역의 정량 + 정성 증거 한 번에 보임.


## 결론 (VectorRAG)

- **① 차원·모델 일치**: `vector(1536)` 컬럼 == 모델 차원(1536) == 행 stamp dimension == 쿼리 임베딩 길이. ✅
- **② count 일치**: 37 chunk == 37 rows. ✅
- **③ 유사도 검색**: golden 15문항 Hit@1/@3·MRR 로 랭킹 정확도 확인. 절대 코사인은 한국어 질의↔영어 문서 cross-lingual 갭으로 0.7 미만이 흔하나, 영어 질의로는 0.7을 넘겨 갭을 정량 확인. 검색 품질 판단은 랭킹 지표 우선.
- **④ sparse baseline 비교**: BM25 (k1=1.5, b=0.75) 대비 Hit@1 +33.3%p, Hit@3 +26.7%p, MRR +0.264. dense embedding 도입의 정량 정당화 근거.
- **⑤ 실제 top-K 살펴보기**: 대표 3문항으로 sparse 와 dense 의 retrieve 결과를 직접 비교. `q-010` 은 sparse 가 score=0.00 으로 완전 실패하는 반면 dense (bge-m3) 가 Hit@1 로 회수 — cross-lingual + 어휘 다양성 케이스의 정성 증거.
